### ISSUE TITLE
Whitespace is not a preprocessing token

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/draft#4766]
Consider [cpp.stringize]
> Let the stringizing argument be the **preprocessing token sequence** for the corresponding argument with placemarker tokens removed. Each occurrence of whitespace between the stringizing argument's preprocessing tokens becomes a single space character in the character string literal. 

Since whitespace is not a preprocessing token, the **preprocessing token sequence** should only consist of each preprocessing token that appears in the corresponding argument. This point is also highlighted in the answer of this issue 
https://stackoverflow.com/a/37462188/11796722

**Comment by jensmaurer** [comment]
I agree that talking about a preprocessing token sequence already disregards whitespace (it's a sequence of tokens, after all), so it makes little sense to talk about the whitespace between preprocessing tokens in the sequence.

What do you envision should be done here editorially?

**Comment by xmh0511** [comment]
Consider this special case
````cpp
#define fun(x)  #x
fun(&123) 
````
`fun(&123)` is expanded to `"&123"`, although the sequence comprises two preprocessing-tokens. Hence, the whitespace is significant here. Since the definition of the stringizing arguments depend on that of arguments of a function-like macro where the arguments list is defined as that

> The **sequence of preprocessing tokens** bounded by the outside-most matching parentheses **forms** the list of arguments for the function-like macro.  

It still does not comprise any whitespace. So, I would like to redefine what is the argument list of a function-like macro. 

> The sequence of preprocessing tokens bounded by the outside-most matching parentheses together with inserting a space into two adjacent preprocessing tokens if there is at least a whitespace between the two preprocessing tokens in the sequence of characters from which the sequence is taken, which forms the sequence of characters of the list of arguments for the function-like macro. The sequences of characters of individual arguments within the list are separated by comma preprocessing tokens, remove whitespace before the first preprocessing token and after the last preprocessing token in the sequence of characters of an argument forms an argument.

Since the effect of the whitespace is used to separate preprocessing tokens, whatever the number of whitespace between them in the original sequence of characters, it seems do not contribute to any case defined in [cpp], except that whether the existence of whitespace between them impacts the result of stringizing.   


[cpp.replace#cpp.stringize-2] could be modified to  
> Let the stringizing argument be the corresponding argument with placemarker tokens removed.  Each occurrence of whitespace between the stringizing argument's preprocessing tokens becomes a single space character in the character string literal.


For instance
````cpp
#define test(x,y) #x#y
````

For macro call  `test(   &123       ,         c )`, its sequence of characters of the list of the arguments is `&123  , c` where the arguments are `&123` and `c`.  

For macro call `test(   &      123,     c  )`, its sequence of characters of the list of the arguments is `& 123, c ` where the arguments are `& 123` and `c`.

**Comment by jensmaurer** [comment]
I agree that multiple whitespace is conflated to a single space in all cases.

However, your examples (in particular the second one) show that the definition and formation of "preprocessing token sequence" itself is broken, and a fix should be applied there instead of to # and ## in isolation. Obviously, a "preprocessing token sequence" is intended to consist of preprocessing tokens, optionally separated by a single space. (Of course, the presence of absence of space only matters for # and ##.)

Given that any work in this area appears to change the specified meaning of ##, this looks non-editorial to me.

**Comment by jensmaurer** [comment]
Regarding [cpp.concat] p2, I think the specification we have is sufficient: p2 says that the macro argument preprocessing token sequence is injected, and p3 says that only the single preprocessing tokens immediately preceding and following ## are spliced into a single preprocessing token. The rest of the tokens arriving by macro argument substitution survive unharmed; that's why you see "1 23" in the output (first example for cpp.concat above).

**Comment by xmh0511** [comment]
[cpp.concat] p2 is not fine for the second example for cpp.concat above. "a parameter is immediately preceded or followed by a ## preprocessing token" sound like a condition, only the condition is satisfied will the parameter be replaced by the corresponding argument. Hence, I think we should remove that condition.

**Comment by jensmaurer** [comment]
http://lists.isocpp.org/core/2021/07/11276.php
